src/services/services.ts

```diff
@@ -2478,6 +2486,58 @@ export function createLanguageService(
         }
     }
 
+    function getJsxLinkedEditAtPosition(fileName: string, position: number): JsxLinkedEditInfo | undefined {
+        const sourceFile = syntaxTreeCache.getCurrentSourceFile(fileName);
+        const token = findPrecedingToken(position, sourceFile);
+        if (!token) return undefined;
+
+        let wordPattern: string | undefined;
+        let ranges: {start: number, end: number}[] = [];
+
+        if ((isJsxOpeningFragment(token.parent) && token.kind===SyntaxKind.LessThanToken)
+            || (isJsxClosingFragment(token.parent) && token.kind===SyntaxKind.SlashToken)){
+
+            const openPos = token.parent.parent.openingFragment.pos + 1;
+            const closePos = token.parent.parent.closingFragment.pos + 2;
+
+            //TODO: fragments with whitespace?
+            ranges = ranges.concat({ start : openPos, end : openPos });
+            ranges = ranges.concat({ start : closePos, end : closePos });
+            wordPattern = undefined;
+        }
+        else if (isJsxFragment(token.parent.parent)) {
+            return undefined;
+        }
+        else {
+            const tag = findAncestor(token,
+                (n) => {
+                    if (!n.parent.parent) return "quit";
+                    else if (isJsxElement(n.parent.parent)) {
+                        if ((isJSXTagName(n) && !isJsxSelfClosingElement(n.parent))
+                            || (n.kind===SyntaxKind.LessThanToken && isJSXTagName(findNextToken(n,n.parent,sourceFile) ?? n))
+                            || n.kind===SyntaxKind.SlashToken) {
+                            return true;
+                        }
+                        return "quit";
+                    }
+                    return false;
+                }
+                )?.parent as JsxOpeningElement | JsxClosingElement | undefined;

```

Comment:
This is super hard to read - can you move the result of `findAncestor` to its own variable?